### PR TITLE
Steinsaltz portal

### PR DESCRIPTION
**Pull request by nsantacruz**
Main changes
- add descriptions to topic author pages
- add portal data model with rendering on topic pages
- refactor newsletter JS component so that it can connect to different newsletter APIs

### CODE REVIEW

File: sefaria/urls.py [hunk 1]
```diff
@@ -385,7 +390,8 @@
 
 # Email Subscribe
 urlpatterns += [
-    url(r'^api/subscribe/(?P<email>.+)$', sefaria_views.subscribe),
+    url(r'^api/subscribe/steinsaltz/(?P<email>.+)$', sefaria_views.subscribe_steinsaltz_newsletter_view),
```
Comment (edamboritz) [inline sefaria/urls.py:393]: I think we can probably generalize this. I dont think we need to have a dedicated view specifically for Steinsaltz.  We can have a general view for a foreign newsletter and perform the routing based on parameter to the correct logic if needed.

File: sefaria/views.py [hunk 2]
```diff
@@ -170,28 +170,62 @@ def accounts(request):
...
+    """
+    API for subscribing to mailing lists, in `lists` url param.
+    Currently active lists are:
+    "Announcements_General", "Announcements_General_Hebrew", "Announcements_Edu", "Announcements_Edu_Hebrew"
+    """
+    body = json.loads(request.body)
+    language = body.get("language", "")
+    educator = body.get("educator", False)
+    crm_mediator = CrmMediator()
+    return crm_mediator.subscribe_to_lists(email, first_name, last_name, educator=educator, lang=language)
```
Comment (edamboritz) [inline sefaria/views.py:209]: I'm not sure why this removes the (not so much) more informative logging + json response there was before?

File: sefaria/model/abstract.py [hunk 3]
```diff
@@ -637,3 +643,71 @@ def foo(obj, **kwargs):
             rec.save()
 
     return foo
+
+class SchemaValidationException(Exception):
```
Comment (edamboritz) [inline sefaria/model/abstract.py:647]: Why are these here and not in the exceptions module?

File: sefaria/model/portal.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+from urllib.parse import urlparse
+from . import abstract as abst
+import urllib.parse
+from django.core.validators import URLValidator
+from django.core.exceptions import ValidationError
+import structlog
+logger = structlog.get_logger(__name__)
+
+
+class InvalidURLException(Exception):
```
Comment (edamboritz) [inline sefaria/model/portal.py:11]: Also why not in exceptions module?

File: sefaria/model/portal.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            "title": ({"en": (str, "required"), "he": (str, "required")}, "required"),
+            "title_url": (str, "optional"),
+            "description": ({"en": (str, "required"), "he": (str, "required")}, "optional"),
+            "api_schema": ({"http_method": (str, "required"),
+                            "payload": ({"first_name_key": (str, "optional"), "last_name_key": (str, "optional"), "email_key": (str, "optional")}, "optional")}
+                           , "optional")
+        }
+
+        if hasattr(self, "about"):
+            abst.validate_dictionary(self.about, about_schema)
```
Comment (edamboritz) [inline sefaria/model/portal.py:95]: Why make them a module level method if you just need to use these validators as inherited methods?

File: sefaria/model/topic.py [hunk 6]
```diff
@@ -69,6 +71,8 @@ def _validate(self):
         if getattr(self, 'subclass', False):
             assert self.subclass in self.subclass_map, f"Field `subclass` set to {self.subclass} which is not one of the valid subclass keys in `Topic.subclass_map`. Valid keys are {', '.join(self.subclass_map.keys())}"
 
+        Portal.validate_slug_exists(self.portal_slug)
```
Comment (edamboritz) [inline sefaria/model/topic.py:74]: Should this not only trigger if there _is_ a portal slug?

**Review by edamboritz — APPROVED** [review]
This looks great! Do take note of a few minor comments though

File: sefaria/model/abstract.py [hunk 7]
```diff
@@ -637,3 +643,71 @@ def foo(obj, **kwargs):
...
+        super().__init__(self.message)
+
+class SchemaInvalidKeyException(Exception):
+    def __init__(self, key):
+        self.key = key
+        self.message = f"Invalid key '{key}' found in data dictionary."
+        super().__init__(self.message)
+
+
+def validate_dictionary(data, schema):
```
Comment (nsantacruz) [inline sefaria/model/abstract.py:666]: @edamboritz I moved this to a new file called `validators.py`

File: sefaria/model/portal.py (hunk 4 above)
Comment (nsantacruz) [reply sefaria/model/portal.py:11]: moved

File: sefaria/model/portal.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class InvalidURLException(Exception):
+    def __init__(self, url):
+        self.url = url
+        self.message = f"'{url}' is not a valid URL."
+        super().__init__(self.message)
+
+
+def validate_url(url):
```
Comment (nsantacruz) [inline sefaria/model/portal.py:18]: @edamboritz I moved this to a new file called `validators.py`

File: sefaria/model/portal.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+
+class InvalidHTTPMethodException(Exception):
+    def __init__(self, method):
+        self.method = method
+        self.message = f"'{method}' is not a valid HTTP API method."
+        super().__init__(self.message)
+
+
+def validate_http_method(method):
```
Comment (nsantacruz) [inline sefaria/model/portal.py:37]: @edamboritz I moved this to a new file called `validators.py`

File: sefaria/model/portal.py (hunk 5 above)
Comment (nsantacruz) [reply sefaria/model/portal.py:108]: This method feels generic enough that it can apply to other classes besides AbstractMongoRecord.

File: sefaria/views.py (hunk 2 above)
Comment (nsantacruz) [reply sefaria/views.py:209]: It doesn't, the logging now happens in `generic_subscribe_to_newsletter_api()`

File: sefaria/urls.py (hunk 1 above)
Comment (nsantacruz) [reply sefaria/urls.py:393]: Did you have in mind that also the Sefaria newsletter would be generalized so there's only one subscribe view?

Comment (edamboritz) [reply sefaria/urls.py:393]: Not necessarily, but it can be (i.e. waht the view defaults to if no parameter is passed in)

File: sefaria/model/portal.py (hunk 5 above)
Comment (edamboritz) [reply sefaria/model/portal.py]: But then it doesnt get hooked into the normal flow of the validate methods on the object?

**Review by edamboritz — APPROVED** [review]
